File: src/main/java/frc/utils/controlWrappers/ProfiledPID.java
package frc.utils.controlWrappers;

import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.trajectory.TrapezoidProfile.Constraints;

/*
 * wrapper for ProfiledPIDController that allows setting gains from a record
 */
public class ProfiledPID extends ProfiledPIDController{
    public ProfiledPID(double kP, double kI, double kD, Constraints constraints){
        super(kP, kI, kD, constraints);
    }
    public ProfiledPID(double kP, double kI, double kD, Constraints constraints, double period){
        super(kP, kI, kD, constraints, period);
    }
    public ProfiledPID(PIDGains.ProfiledPID gains){
        super(gains.kP, gains.kI, gains.kD, new Constraints(gains.maxSpeed, gains.maxAccel));
    }

    public void setSpeed(double maxSpeed){
        super.setConstraints(new Constraints(maxSpeed, super.getConstraints().maxAcceleration));
    }
    public void setAccel(double maxAccel){
        super.setConstraints(new Constraints(super.getConstraints().maxVelocity, maxAccel));
    }
    public double getMaxSpeed(){
        return super.getConstraints().maxVelocity;
    }
    public double getMaxAccel(){
        return super.getConstraints().maxAcceleration;
    }
    public void setGains(PIDGains.ProfiledPID gains){
        setPID(gains.kP, gains.kI, gains.kD);
        setConstraints(new Constraints(gains.maxSpeed, gains.maxAccel));
    }
}

File: src/main/kotlin/frc/utils/controlWrappers/ProfiledPID.kt
package frc.utils.controlWrappers

import edu.wpi.first.math.controller.ProfiledPIDController
import edu.wpi.first.math.trajectory.TrapezoidProfile.Constraints

/*
 * wrapper for ProfiledPIDController that allows setting gains from a record
 */
class ProfiledPID : ProfiledPIDController {
    constructor(kP: Double, kI: Double, kD: Double, constraints: Constraints) : super(kP, kI, kD, constraints)
    constructor(kP: Double, kI: Double, kD: Double, constraints: Constraints, period: Double) : super(kP, kI, kD, constraints, period)
    constructor(gains: PIDGains.ProfiledPID) : super(gains.kP, gains.kI, gains.kD, Constraints(gains.maxSpeed, gains.maxAccel))

    fun setSpeed(maxSpeed: Double) {
        setConstraints(Constraints(maxSpeed, super.getConstraints().maxAcceleration))
    }

    fun setAccel(maxAccel: Double) {
        setConstraints(Constraints(super.getConstraints().maxVelocity, maxAccel))
    }

    fun getMaxSpeed(): Double {
        return super.getConstraints().maxVelocity
    }

    fun getMaxAccel(): Double {
        return super.getConstraints().maxAcceleration
    }

    fun setGains(gains: PIDGains.ProfiledPID) {
        setPID(gains.kP, gains.kI, gains.kD)
        setConstraints(Constraints(gains.maxSpeed, gains.maxAccel))
    }
}
